Edge curvature

ekkis opened the pull request:
#273

This patch enhances the library’s edge curvature capabilities, which
are at present hard-coded.

Of particular interest are cases when multiple edges exist between two
nodes.  the current framework displays edges atop one another so what
is needed is a way to specify curvature for individual edges.

the patch to sigma.utils.getQuadraticControlPoint() adds a fifth
parameter which contains a collection of coefficients dictating the
control point’s position.

patches to all calling contexts have been made to provide the
coefficients as found on the edge, or to pick up the global settings
value ’curvatureCoefficients’.  if neither exists the current default
is used.

additionally, a method sigma.canvas.edges.autoCurve() has been provided
to automatically curve edges for nodes with multiple connections.  this
method is accessible as an overload to the sigma.canvas.edges.curve()
method.

finally, the parallelEdges module overrides the functionality of the
getQuadraticControlPoint() function which is a pernicious approach to
solve the same problem and has been re-implemented as
getQuadraticControlPoint2 by calling the original method, with all
calls to it from the same module patched

sheymann commented:
Thanks! Can you merge `develop` into your branch in order to get a clean diff please?

```sh
git checkout develop
git pull
git checkout edge-curvature
git merge develop
```

ekkis commented:
I tried but got an error:

aprilis:linkurious.js ekkis$ git checkout develop
error: pathspec 'develop' did not match any file(s) known to git.

but my local copy comes from my own fork at: https://github.com/ekkis/linkurious.js <https://github.com/ekkis/linkurious.js>

> 24 sep 2015 kl. 00:14 skrev Sébastien Heymann <notifications@github.com>:
> 
> Thanks! Can you merge develop into your branch in order to get a clean diff please?
> 
> git checkout develop
> git pull
> git checkout edge-curvature
> git merge develop
> —
> Reply to this email directly or view it on GitHub <https://github.com/Linkurious/linkurious.js/pull/272#issuecomment-142836627>.
>

ekkis commented:
to be more helpful:

aprilis:linkurious.js ekkis$ git remote -v
Linkurious	https://github.com/Linkurious/linkurious.js.git (fetch)
Linkurious	https://github.com/Linkurious/linkurious.js.git (push)
origin	https://github.com/ekkis/linkurious.js.git (fetch)
origin	https://github.com/ekkis/linkurious.js.git (push)

and what I think you wanted me to do:

aprilis:linkurious.js ekkis$ git checkout linkurious-version
Switched to branch 'linkurious-version'
Your branch and 'origin/linkurious-version' have diverged,
and have 72 and 5 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

aprilis:linkurious.js ekkis$ git merge Linkurious/linkurious-version
Removing CONTRIBUTING.draft.md
Merge made by the 'recursive' strategy.
 CONTRIBUTING.draft.md | 47 -----------------------------------------------
 1 file changed, 47 deletions(-)
 delete mode 100644 CONTRIBUTING.draft.md

aprilis:linkurious.js ekkis$ git checkout edge-curvature
Switched to branch 'edge-curvature'
Your branch is up-to-date with 'origin/edge-curvature'.

aprilis:linkurious.js ekkis$ git merge linkurious-version
Removing CONTRIBUTING.draft.md
Merge made by the 'recursive' strategy.
 CONTRIBUTING.draft.md | 47 -----------------------------------------------
 1 file changed, 47 deletions(-)
 delete mode 100644 CONTRIBUTING.draft.md

so now what?


> 24 sep 2015 kl. 13:44 skrev Erick Calder <e@arix.com>:
> 
> I tried but got an error:
> 
> aprilis:linkurious.js ekkis$ git checkout develop
> error: pathspec 'develop' did not match any file(s) known to git.
> 
> but my local copy comes from my own fork at: https://github.com/ekkis/linkurious.js <https://github.com/ekkis/linkurious.js>
> 
>> 24 sep 2015 kl. 00:14 skrev Sébastien Heymann <notifications@github.com <mailto:notifications@github.com>>:
>> 
>> Thanks! Can you merge develop into your branch in order to get a clean diff please?
>> 
>> git checkout develop
>> git pull
>> git checkout edge-curvature
>> git merge develop
>> —
>> Reply to this email directly or view it on GitHub <https://github.com/Linkurious/linkurious.js/pull/272#issuecomment-142836627>.
>> 
>

ekkis commented:
ok, after the above, I tried again:

```
aprilis:linkurious.js ekkis$ git checkout develop
Branch develop set up to track remote branch develop from Linkurious.
Switched to a new branch 'develop'
aprilis:linkurious.js ekkis$ git pull
Already up-to-date.
aprilis:linkurious.js ekkis$ git checkout edge-curvature
Switched to branch 'edge-curvature'
Your branch is ahead of 'origin/edge-curvature' by 3 commits.
  (use "git push" to publish your local commits)
aprilis:linkurious.js ekkis$ git merge develop
Auto-merging src/utils/sigma.utils.js
Auto-merging src/sigma.core.js
Auto-merging src/misc/sigma.misc.bindEvents.js
CONFLICT (modify/delete): src/classes/sigma.classes.edgequad.js deleted in develop and modified in HEAD. Version HEAD of src/classes/sigma.classes.edgequad.js left in tree.
Auto-merging plugins/sigma.renderers.linkurious/canvas/sigma.canvas.nodes.def.js
CONFLICT (content): Merge conflict in plugins/sigma.renderers.linkurious/canvas/sigma.canvas.nodes.def.js
Auto-merging plugins/sigma.renderers.linkurious/canvas/sigma.canvas.edgehovers.curvedArrow.js
Auto-merging plugins/sigma.renderers.linkurious/README.md
CONFLICT (content): Merge conflict in plugins/sigma.renderers.linkurious/README.md
Auto-merging plugins/sigma.renderers.halo/sigma.renderers.halo.js
Auto-merging plugins/sigma.renderers.edgeLabels/sigma.canvas.edges.labels.curve.js
Removing plugins/sigma.plugins.dragNodes/settings.js
Auto-merging plugins/sigma.parsers.cypher/sigma.parsers.cypher.js
CONFLICT (content): Merge conflict in plugins/sigma.parsers.cypher/sigma.parsers.cypher.js
Auto-merging examples/renderers-linkurious-edges-shapes.html
Auto-merging examples/plugin-exporters-json.html
Removing examples/plugin-customShapes.html
Removing examples/plugin-customShapes-icons.html
Removing examples/plugin-customShapes-hybrid.html
Automatic merge failed; fix conflicts and then commit the result.
```

ekkis commented:
after that, I opened up the GitHub OSX GUI client, did a commit (which worked without complaining about having to manually merge, and did a sync.  is there anything else I need to do?

sheymann commented:
It went bad as you commited conflicted files instead of resolving conflicts, see sigma.parsers.cypher.js in https://github.com/Linkurious/linkurious.js/pull/272/files for instance. Please see how to resolve conflicts on http://githowto.com/resolving_conflicts

The probable reason it happens is because you work on the same branch since the last 3 pull requests, instead of branching from `develop` each time you work on a new feature. You can learn how to work with branches here: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

sheymann commented on plugins/sigma.renderers.linkurious/canvas/sigma.canvas.edges.curve.js:
```diff
@@ -105,4 +109,45 @@
+		var step = cc.step(cc.length, n);
+		if (count[key].n > 1) {
+			this.type = cc.type;
+			this.cc = $.extend({}, cc, cc.calc(cc.length, step, count[key].i++));
```
`$` is not available in sigma. Please replace `$.extend` by `sigma.utils.extend`.

sheymann commented on src/utils/sigma.utils.js:
```diff
@@ -229,12 +229,16 @@
    */
-  sigma.utils.getQuadraticControlPoint = function(x1, y1, x2, y2) {
+  sigma.utils.getQuadraticControlPoint = function(x1, y1, x2, y2, cc) {
+	cc = $.extend({ x: 2, y: 4 }, cc);
```
`$` is not available in sigma. Please replace `$.extend` by `sigma.utils.extend`.

ekkis commented:
oh yes, my bad, I've made the fix

sheymann commented:
sigma.classes.edgequad.js has been removed, please patch sigma.classes.quad.js

sheymann commented:
How about removing the file?

ekkis commented:
better now?

sheymann commented:
@ekkis what is the domain (min,max) of the setting `curvatureCoefficients`?

Can you provide an html example? I cannot make it work.

ekkis commented:
yes.  let me put something together

sheymann commented:
I've also noticed that examples/parallel-edges.html doesn't render parallel edges anymore. Can you fix it please?

sheymann commented:
In examples/events.html, `autoCurve` is never called even if all edges have same source and target nodes.

sheymann commented:
@ekkis any update on this? I cannot move forward.

ekkis commented:
sorry mate.  I've been just buried and haven't had the time to work on this.  I know you're stuck.
